File: test/book_security_test.rb
require "test_helper"
require "bcrypt"

class BookSecurityTest < IntegrationTest
  parallelize_me!

  def test_books_index_redirects_anonymous_users_to_login
    user_id = session.db[:users].insert username: "test", password_digest: BCrypt::Password.create("secret phrase")
    session.db[:books].insert user_id: user_id, title: "Foo", author: "Bar", finish_date: Date.new(2019,1,1), media: "book"

    browser.goto "#{session.app_host}/users/test/books"
    assert_equal "#{session.app_host}/login", browser.url
  end

  def test_books_index_is_forbidden_to_other_users
    user_id = session.db[:users].insert username: "test", password_digest: BCrypt::Password.create("secret phrase")
    session.db[:books].insert user_id: user_id, title: "Foo", author: "Bar", finish_date: Date.new(2019,1,1), media: "book"

    other_user_id = session.db[:users].insert username: "other", password_digest: BCrypt::Password.create("secret phrase")

    browser.goto "#{session.app_host}/login"
    browser.text_field(label: "Username").set "other"
    browser.text_field(label: "Password").set "secret phrase"
    browser.button(text: "Login").click
    assert browser.button(text: "Logout").exist?

    browser.goto "#{session.app_host}/users/test/books"
    assert_match /Forbidden/, browser.text
  end
end

File: test/browser/book_security_test.go
package browser_test

import (
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/jackc/booklog/test/testdata"
	"github.com/stretchr/testify/require"
)

func TestBooksIndexRedirectsAnonymouseUsersToLogin(t *testing.T) {
	t.Parallel()

	ctx := context.Background()
	serverInstance := startServer(t)
	db := serverInstance.DB.Connect(t, ctx)
	page := TestBrowserManager.Acquire(t).Page()

	user := testdata.CreateUser(t, db, ctx, map[string]any{"username": "test", "password": "secret phrase"})
	testdata.CreateBook(t, db, ctx, map[string]any{"user_id": user["id"], "title": "Foo", "author": "Bar", "finish_date": time.Date(2019, 1, 1, 0, 0, 0, 0, time.Local), "format": "text"})
	page.MustNavigate(fmt.Sprintf("%s/users/test/books", serverInstance.Server.URL))
	page.HasContent("form", "Login")
	require.Equal(t, fmt.Sprintf("%s/login", serverInstance.Server.URL), page.MustInfo().URL)
}

func TestBooksIndexIsForbiddenToOtherUsers(t *testing.T) {
	t.Parallel()

	ctx := context.Background()
	serverInstance := startServer(t)
	db := serverInstance.DB.Connect(t, ctx)
	page := TestBrowserManager.Acquire(t).Page()

	user := testdata.CreateUser(t, db, ctx, map[string]any{"username": "test", "password": "secret phrase"})
	testdata.CreateBook(t, db, ctx, map[string]any{"user_id": user["id"], "title": "Foo", "author": "Bar", "finish_date": time.Date(2019, 1, 1, 0, 0, 0, 0, time.Local), "format": "text"})

	testdata.CreateUser(t, db, ctx, map[string]any{"username": "other", "password": "secret phrase"})
	login(t, ctx, page, serverInstance.Server.URL, "other", "secret phrase")
	page.MustNavigate(fmt.Sprintf("%s/users/test/books", serverInstance.Server.URL))
	page.HasContent("body", "Forbidden")
}
